Completely rewrite `Code` mobject + add support for short notation RGB(A) hex code specifications to `ManimColor`

behackl opened the pull request:
See title.

Due to a version upgrade of `pygments`, support for short notation color hex codes was required -- which also required a patch to `Code`. This part of this PR could be factored out (e.g., to allow us to finalize the release faster).

The `Code` mobject was in a somewhat bad state, code quality wise. This re-implementation is clean(er) and easier to maintain / extend (especially considering the interest in terms of adding specific transform and highlight animations in recent PRs and issues).

**Note:** I did not keep the interface of the old mobject -- in fact, there are several changes to the names of the keyword arguments etc., and I've also not ported all of the functionality. I'm open to porting over further features (like configureable tab width etc.) though.

---

Summary of breaking changes for the `Code` class, by keyword arguments. Old arguments:

- `file_name` -- renamed to `code_file`
- `code` -- renamed to `code_string`
- `tab_width` -- kept as is
- `line_spacing` -- kept, but has to be passed inside of `paragraph_config` dict now
- `font_size` -- kept, but has to be passed inside of `paragraph_config` dict now
- `font` -- kept, but has to be passed inside of `paragraph_config` dict now
- `stroke_width`  -- kept, but has to be passed inside of `paragraph_config` dict now
- `margin` -- renamed to `buff` and has to be passed inside of `background_config` dict now
- `indentation_chars` -- gone
- `background` -- kept as is
- `background_stroke_width` -- renamed to `stroke_width` and has to be passed inside of `background_config` dict now
- `background_stroke_color` -- renamed to `stroke_color` and has to be passed inside of `background_config` dict now
- `corner_radius` -- kept, but has to be passed inside of `background_config` dict now
- `insert_line_no` -- renamed to `add_line_numbers`
- `line_no_from` -- renamed to `line_numbers_from`
- `line_no_buff` -- gone
- `style` -- renamed to `formatter_style`
- `language` -- kept as is
- `generate_html_file` -- gone
- `warn_missing_font` -- gone
- `**kwargs` -- gone 

New keyword arguments:

- `background_config` -- extra arguments passed to background mobject (currently: `SurroundingRectangle`), default arguments are stored in class attribute `Code.default_background_config`
- `paragraph_config` -- extra arguments passed to `Paragraph` (for code and line numbers), default arguments are stored in class attribute `Code.default_paragraph_config`.

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -6,50 +6,24 @@
 
-import numpy as np
-from pygments import highlight, styles
+from bs4 import BeautifulSoup
```
Should we add this as an explicit dependency?

behackl replied:
For sure, I just forgot about it: 4c3ea2f7

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
-        self.move_to(np.array([0, 0, 0]))
-
-    @classmethod
-    def get_styles_list(cls):
```
This method seems like it would be useful to have?

behackl replied:
Probably a good idea, convenience-wise. Added back in 3a45b230.

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
-        line_no_from: int = 1,
-        line_no_buff: float = 0.4,
-        style: str = "vim",
+        code_file: os.PathLike | str | None = None,
```
We have [`manim.typing.StrPath`](https://docs.manim.community/en/latest/reference/manim.typing.html#manim.typing.StrPath)

behackl replied:
Thanks! 2de9adca

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        formatter_style: str = "vim",
+        line_numbers: bool = True,
+        line_numbers_from: int = 1,
+        background: str = "rectangle",
```
```suggestion
        background: Literal["rectangle", "window"] = "rectangle",
```
Or better yet, an `Enum` might work better.

behackl replied:
Kept as a `Literal` for now to keep things slightly simpler, but I'm open to introducing a `CodeBackgroundType(Enum)`. 815b9dc4

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        line_numbers_from: int = 1,
+        background: str = "rectangle",
+        background_margin: float = 0.3,
+        background_stroke_color: str = "#fff",
```
Maybe we want to use [`ParsableManimColor`](https://docs.manim.community/en/latest/reference/manim.utils.color.core.html#manim.utils.color.core.ParsableManimColor)?
Also using `WHITE` might be better than the raw hex.

behackl replied:
Agreed! 9998ca46

JasonGrace2282 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        if line_numbers:
+            paragraph_kwargs.update(
+                {
+                    "alignment": "right",
```
Minor nitpick to get it to format better:
```suggestion
                    "alignment": "right"
```
so that it collapses it into one line

behackl replied:
True! e1fb43a5

JasonGrace2282 reviewed (changes requested):
Haven't tested this yet, but just some stuff I saw from looking through.

On another note, since the `Code` mobject has been fully rewritten, it might be a good idea to enable `mypy` on it (see `mypy.ini` in the root directory).

chopan050 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        soup = BeautifulSoup(
+            highlight(code_string, lexer, formatter), features="html.parser"
+        )
+        self._code_html = soup.find("pre")
```
In general, `BeautifulSoup.find()` returns `Tag | NavigableString | None`. Only `Tag` has a `children` attribute. To make MyPy happy, we could add an assertion:

```suggestion
        self._code_html = soup.find("pre")
        assert isinstance(self._code_html, Tag)
```

behackl replied:
LGTM!

chopan050 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        color_ranges = []
+        current_line_color_ranges = []
+        current_line_char_index = 0
+        for child in self._code_html.children:
+            if child.name == "span":
+                try:
+                    color = child["style"].removeprefix("color: ")
```
`child` doesn't have a `name` attribute unless it's also a `Tag`, which is implicitly given. We should also assert that `child` is a `Tag`, either here or along with the previous assertion.

`child["style"]` can be either a string or a list of strings which does not have `.removeprefix()`. Maybe we should also assert that `child["style"]` is a `str`.

The code does get a bit clunky, so any suggestions are accepted.

```suggestion
        for child in self._code_html.children:
            assert isinstance(child, Tag)
            if child.name == "span":
                try:
                    style = child["style"]
                    assert isinstance(style, str)
                    color = style.removeprefix("color: ")
```

behackl replied:
`child` can also be `NavigableString`, the assertion will have to be rewritten slightly. Otherwise a good idea though, 👍

chopan050 replied:
I just checked, and indeed: it can be a `NavigableString`. So yeah, it would be `assert isinstance(child, (Tag, NavigableString))`.

behackl replied:
Actually, `self._code_html.children` is already typed as `Iterator[PageElement]`, that seems strong enough to me. I've strengthened the typing within the `try`-block in d110d5ec

chopan050 reviewed (changes requested):
Thanks for the very needed rewriting of `Code` and the support of shorter hex code colors!

I left some comments regarding `Code`.

behackl commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        font: str = "Monospace",
+        font_size: float = 24.0,
+        line_spacing: float = 0.5,
+        **paragraph_kwargs,
```
I am not sure two separate config dicts for code lines and line numbers are really necessary. If the line spacing and font size do not match, the mobject will look very broken. If users want to modify one of the two `Paragraph` objects, I'd rather recommend accessing the mobjects directly via `code_mobject.line_numbers` or `code_mobject.code_lines`.

I don't mind turning the catchall-kwarg into a single `dict[str, Any]` argument though, will do!

behackl replied:
rewritten: 32ce1521

behackl commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
-        strati = self.html_string.find("background:")
-        self.background_color = self.html_string[strati + 12 : strati + 19]
-        self._gen_code_json()
+        if code_file is None and code_string is None:
+            raise ValueError("Either a code file or a code string must be specified.")
```
Seems fine to me, will apply this change after locally consulting mypy too. 👍

behackl replied:
done in a7ad59a9

chopan050 commented on manim/mobject/text/code_mobject.py:
```diff
@@ -81,557 +59,207 @@ def construct(self):
-        **kwargs,
+        formatter_style: str = "vim",
+        tab_width: int = 4,
+        line_numbers: bool = True,
```
The name should reflect that this is, indeed, a boolean. Maybe something like `add_line_numbers` or `use_line_numbers`?
```suggestion
        add_line_numbers: bool = True,
```

behackl replied:
Agreed! Went with `add_line_numbers` -- f3692312

behackl commented on manim/mobject/text/code_mobject.py:
```diff
@@ -89,549 +64,186 @@ def construct(self):
+        background_margin: float = 0.3,
+        background_stroke_color: str = "#fff",
+        background_fill_color: str = "#222",
+        background_corner_radius: float = 0.2,
```
rewritten: dfba6661

I think I'd like to keep the background implementation as-is for now, in case someone wants to extend the mobject with other background types, then a boolean switch would be less flexible again.

behackl commented:
Thanks for your reviews! I've addressed all suggestions, the interface of the rewritten mobject now looks quite different than in the first version -- but I think it was improved significantly and brought in line to how other mobjects composed of different sub-mobjects like `Axes` are constructed.

I've updated the overview in the PR description, the main takeaway is that there are now two `dict[str, Any]` arguments `paragraph_config` and `background_config` to which further arguments for `Paragraph` and the `SurroundingRectangle` can be passed. Semantically, the implementation could be improved somewhat by renaming these `paragraph_config_extra` and `background_config_extra`, but I also don't think this is bad as it is currently.

JasonGrace2282 reviewed (approved):
LGTM!
I noticed we don't have any graphical tests for `Code` - maybe in this PR or a followup, it might be a good idea to add one?

behackl commented:
> LGTM! I noticed we don't have any graphical tests for `Code` - maybe in this PR or a followup, it might be a good idea to add one?

IIRC the issue with graphical tests was that `Text` and `Paragraph` do not render pixel-consistently across systems, but it might be worth to look into this again -- if needs be, perhaps with OS-dependent control data.

chopan050 reviewed (approved):
LGTM!
